pkg/deploy/local/local_provider.go

```diff
@@ -20,98 +26,166 @@ import (
+// applies the filter to all selected workloads in selected namespaces
+func (p *Runner) RunFilter(filter *v1.FilterSpec) error {
+	cfg, err := p.getConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if err := p.Input.Close(); err != nil {
-		return nil, err
+	image, err := p.Store.Get(filter.Image)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve image. make sure to run `wasme pull %v` to pull the image to your local storage.", filter.Image)
 	}
-
-	if !p.UseJsonConfig {
-		var err error
-		b, err = yaml.YAMLToJSON(b)
+	if filter.RootID == "" {
+		imageCfg, err := image.FetchConfig(p.Ctx)
 		if err != nil {
-			return nil, err
+			return err
 		}
+		roots := imageCfg.GetConfig().GetRootIds()
+		if len(roots) == 0 {
+			return errors.Errorf("found no root_id on image or in params")
+		}
+
+		// default to first root
+		filter.RootID = roots[0]
 	}
 
-	var bootstrap envoy_config_bootstrap_v2.Bootstrap
-	return &bootstrap, wasmeutil.UnmarshalBytes(b, &bootstrap)
-}
+	// allow filter ID to be empty, as we don't care in local envoy
+	if filter.Id == "" {
+		filter.Id = filter.RootID
+	}
 
-func (p *Provider) writeConfig(bootstrap *envoy_config_bootstrap_v2.Bootstrap) error {
-	b, err := wasmeutil.MarshalBytes(bootstrap)
+	filterDir, err := p.Store.Dir(filter.Image)
 	if err != nil {
 		return err
 	}
-
-	if !p.UseJsonConfig {
-		b, err = yaml.JSONToYAML(b)
-		if err != nil {
-			return err
-		}
+	filterDir, err = filepath.Abs(filterDir)
+	if err != nil {
+		return err
 	}
+	filterFile := filepath.Join(filterDir, model.CodeFilename)
 
-	var out io.Writer
-	if p.OutFile == "-" {
-		// use stdout
-		out = os.Stdout
-	} else {
-		f, err := os.Create(p.OutFile)
-		if err != nil {
-			return err
-		}
-		out = f
+	if err := addFilterToListeners(filter, cfg.GetStaticResources().GetListeners(), filterFile); err != nil {
+		return err
 	}
-	_, err = out.Write(b)
-	return err
-}
 
-// applies the filter to all selected workloads in selected namespaces
-func (p *Provider) ApplyFilter(filter *v1.FilterSpec) error {
-	cfg, err := p.getConfig()
+	configYaml, err := marshalConfig(cfg)
 	if err != nil {
 		return err
 	}
 
-	if err := addFilterToListeners(filter, cfg.GetStaticResources().GetListeners(), p.FilterPath); err != nil {
+	if p.Output != nil {
+		_, err = p.Output.Write(configYaml)
 		return err
 	}
 
-	return p.writeConfig(cfg)
-}
+	logrus.Infof("mounting filter file at %v", filterFile)
 
-// removes the filter from all selected workloads in selected namespaces
-func (p *Provider) RemoveFilter(filter *v1.FilterSpec) error {
-	cfg, err := p.getConfig()
+	logrus.Debugf("using bootstrap config: \n%s", string(configYaml))
+
+	ports, err := getListenerPorts(cfg)
 	if err != nil {
 		return err
 	}
 
-	if err := removeFilterFromListeners(filter, cfg.GetStaticResources().GetListeners()); err != nil {
+	dockerArgs := append([]string{
+		"--rm",
+		"--name", filter.Id,
+		"--entrypoint", "envoy",
```

Comment:
i would also add `--disable-hot-restart` here for good measure.

Reply:
see `envoyArgs` below